### ISSUE TITLE
How to download and merge video from "X" and audio from "best" when both files have video and audio?

### ISSUE THREAD

**Issue by kuco28** [yt-dlp/yt-dlp#2228]
### Checklist

- [X] I'm asking a question and **not** reporting a bug/feature request
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar questions including closed ones

### Question

There is no separate audio files. Bigger resolution = better audio.
For example 480p+best download only first. 
--compat-options multistreams give file with two video and audio tracks.
Is it possible to take video from one and audio from another using one command or I have download both and remux using selected tracks for example by ffmpeg?

**Comment by arbarbosa1979** [comment]
Which site are you talking about exactly? Because depending on the site, video and audio come together, so you can't collect audio from one and video from another.

**Comment by pukkandan** [comment]
You have to download both entirely and mux them together

**Comment by kuco28** [comment]
I know I have to download both. I meant muxing selected tracks **automatically** like [youtube-dl ](https://github.com/ytdl-org/youtube-dl)does instead of muxing them manually after download.

**Comment by pukkandan** [comment]
I meant that you have to mux them manually

There is no in-built way to discard certain streams during merge. It *may* be possible to do with `--ppa`, but you will have to experiment with it (I haven't tried)

> I meant muxing selected tracks automatically like youtube-dl does

youtube-dl can't do this either

**Comment by kuco28** [comment]
> youtube-dl can't do this either

I tried it many times before I moved to yt-dlp and it worked so it's possible there.

**Comment by pukkandan** [comment]
Maybe I misunderstood your question. Show the verbose log of youtube-dl doing what you want

**Comment by fstirlitz** [comment]
> you can test it out for example on random YouTube video passing youtube-dl -f 18+22 LINK (both have video and audio) and it will mux it (ffmpeg needed) by taking video from first and audio from second

I’m pretty sure this wasn’t an intended use case of the original upstream project, and it worked only by accident. But in that case, I guess this is a side effect of blackjack4494/youtube-dlc#81 (specifically, <https://github.com/blackjack4494/youtube-dlc/pull/81/commits/d03cfdce1bce28284fa0f168d5d278caacd2fa0b#diff-e77497ca8029f8d381974fd189b343311d8f1323f49ebae2203cfaec91e98f8eL521-R526>), from before this project was even formed.

**Comment by pukkandan** [comment]
I also don't think this was intended. But as Linus once said, a bug that people depend on is a feature. It would be difficult to "fix" this in yt-dlp now for normal operation. But I should be able to make it work atleast with a compat option

**Comment by chrizilla** [comment]
> There is no in-built way to discard certain streams during merge.

Doesn't `-f bestvideo*+bestaudio` discard the `bestvideo*`'s audio stream (if present) when merging with `bestaudio` ?

**Comment by campones** [comment]
on my side I m trying to simply select an average quality format , audio and video being split but it just download the video.
If I don't select format then it's working fine by downloading the best quality audio and video then merging the both

yt-dlp https://www.france.tv/france-5/c-dans-l-air/3015299-covid-scandales-boris-johnson-contre-attaque.html -f hls-1582-1+hls-1582-0

any ideas?

**Comment by MrRawes** [comment]
> > @pukkandan [said](https://github.com/yt-dlp/yt-dlp/issues/2228#issuecomment-1005599488): There is no in-built way to discard certain streams during merge.
> 
> Doesn't `-f bestvideo*+bestaudio` discard the `bestvideo*`'s audio stream (if present) when merging with `bestaudio` ?

no bestvideo*+bestaudio selects best video source even with audio and if there is not audio pressent it gets the best audio source

**Comment by ScriptCitty** [comment]
This took me a while to figure out but some f3 in the read me did some justice.
I have this. Working on windows with ffmpeg installed.
yt-dlp --audio-multistreams -S ext -P "C:\path" "videoLink"

--audio-multistreams            Allow multiple audio streams to be merged into a single file.
This is because everyone knows 1080p youtube has no audio.
If you don't do this, you'll be having a blank audio clip over a good audio clip.
doing this on particular videos will leave you by default with webm extension.

*[Format Sorting](https://github.com/yt-dlp/yt-dlp/blob/master/README.md#sorting-formats): The default format sorting 
options have been changed so that higher resolution and better codecs will be now preferred instead of simply using larger 
bitrate. Furthermore, you can now specify the sort order using -S. This allows for much easier format selection than what is 
possible by simply using --format

vext: Video Extension (mp4 > mov > webm > flv > other). If --prefer-free-formats is used, webm is preferred.
aext: Audio Extension (m4a > aac > mp3 > ogg > opus > webm > other). If --prefer-free-formats is used, the order changes to ogg > opus > webm > mp3 > m4a > aac
ext: Equivalent to vext,aext


Hope this answers it for you.

**Comment by gnuyen** [comment]
> I also don't think this was intended. But as Linus once said, a bug that people depend on is a feature. It would be difficult to "fix" this in yt-dlp now for normal operation. But I should be able to make it work atleast with a compat option

Is this because people are dependent on the other behavior in various settings?   My use case for this in particular is when the video has a particular language audio attached and the alternate language track is provided as audio only.  For my use case it doesn't make sense to mux them afterwards, because I want to be able to watch live streams, so I am also using "-o -" to pipe it into ffmpeg and/or vlc.

Are we against enabling the arbitrary stream merges in general or are we okay with doing it as long as there is some kind of flag to specify that is what is occurring.   To me it's most reasonable if you explicitly specify a video and a separate audio source, that the expected behavior is to use the video from the first and the audio from the second, not just ignore the second if the first has audio, but if this causes compatibility issues it's understandable to have some kind of flag.

**Comment by pukkandan** [comment]
@gnuyen Your use-case should be covered by `--audio-multistreams`, no? OP's issue was that they specifically want to discard the other streams

**Comment by gnuyen** [comment]
> @gnuyen Your use-case should be covered by `--audio-multistreams`, no? OP's issue was that they specifically want to discard the other streams

I'll check during the next live event, but as far as I could tell using -o - and --audio-multistreams and piping it to ffmpeg didn't seem to work.  I couldn't access the second audio stream at all and didn't see one, but maybe I was using ffmpeg wrong.  I will check tomorrow as this will for sure come up again

**Comment by CirnoChan999** [comment]
Hello,

I want to add that if you want to select a specific video and audio streams (formats) to combine, you can do the following:

1. Get the available audio and video streams
`yt-dlp --list-formats <YOUTUBE LINK>`

2. Download and combine the formats you want to combine using their IDs
`yt-dlp --format <VIDEO FORMAT ID>+<AUDIO FORMAT ID> <YOUTUBE LINK>`

Example:

`yt-dlp --list formats https://www.youtube.com/watch?v=aqz-KE-bpKQ`
`yt-dlp --format 17+599 https://www.youtube.com/watch?v=aqz-KE-bpKQ`

**Comment by Tboiii-123** [comment]
i have an issue, using  only best gives me  low quality video and audio but when using bestvideo  gives me good quality video but without an audio
and when i combined bestvideo +bestaudio am having an error talking about ffmpeg path

ERROR: You have requested merging of multiple formats but ffmpeg is not installed. Aborting due to --abort-on-error
Failed to download https://youtube.com/shorts/nAD-1ffz6yk?si=gq9e4t2BiO-p9uXl: ERROR: You have requested merging of multiple formats but ffmpeg is not installed. Aborting due to --abort-on-error
PS C:\Users\HP\Desktop\Programming\Python_programming\Django Tutorial\Youtube downloader\youtube_d>


Please how can i reslove it i wanna use it for my django project

**Comment by seproDev** [comment]
@Tboiii-123 I don't think that is related to this issue. You are missing ffmpeg. `best` means the best pre-merged format. For YouTube, that means 360p. `bestvideo` means best format that only contains a video. If you have ffmpeg installed, you want to leave everything at the default and yt-dlp will automatically pick what it thinks is the best quality formats. If even after installing ffmpeg you run in to problems, please open a new issue with a complete verbose log.

**Comment by fc1104** [comment]
If I specify the video with 720 resolution and the best audio,' format': f' best video [height < = {schema. height}]+best audio', how can I merge correctly without generating local files? There are various problems when I merge through ffmpeg. The following is the configuration of ydl_opts:
ydl_opts = {
Format': f' best video [height < = {schema. height}]+best audio', # Set the video quality.
Outtmpl': '-', # Do not save the file locally.
Cachedir': False, # Disable caching.
Nopart': True, # Disable staged download.
Nooverwrites': False, # Do not overwrite the existing file.
Quiet': False, # Reduce output
'merge_output_format': 'mp4',
'ffmpeg_location': r'D:\software\ffmpeg-7.0.2-essentials_build\bin\ffmpeg.exe'
}
